File: kdb-document/src/commonMain/kotlin/dev/kdb/document/DocumentTree.kt
package dev.kdb.document

import dev.kdb.codec.KdbHash
import dev.kdb.codec.KdbTimestamp
import dev.kdb.codec.KdbUuid
import dev.kdb.codec.KdbValue
import dev.kdb.codec.toKdbTimestamp
import dev.kdb.codec.toKdbUuid
import dev.kdb.codec.toTimestampVal
import dev.kdb.codec.toUuidVal

/**
 * Materialised doc id → content hash map at a commit.
 *
 * trieRoot is intentionally a body property, not a primary-constructor
 * one: it backs incremental with/without updates (O(delta) via
 * DocumentTreeTrie.kt instead of O(entries.size)) but must NOT affect
 * this data class's generated equals/hashCode/copy/toString, which
 * existing code relies on comparing by treeHash/entries alone. Trees
 * built via build() from a flat map (e.g. wire decode) don't carry a
 * trieRoot - with/without fall back to a one-time O(n) trieBuild in that
 * case, then are incremental from then on. Always correct either way;
 * only the constant differs.
 */
public data class DocumentTree(
    public val treeHash: KdbHash,
    public val entries: Map<KdbUuid, KdbHash>,
) {
    internal var trieRoot: TrieNode? = null

    public val size: Int get() = entries.size

    public fun contains(docId: KdbUuid): Boolean = entries.containsKey(docId)

    public fun hashFor(docId: KdbUuid): KdbHash? = entries[docId]

    public fun with(
        docId: KdbUuid,
        contentHash: KdbHash,
    ): DocumentTree {
        val root = trieInsert(trieRootOrBuild(), docId, contentHash)
        val newEntries = entries + (docId to contentHash)
        return DocumentTree(trieTreeHash(root), newEntries).also { it.trieRoot = root }
    }

    public fun without(docId: KdbUuid): DocumentTree {
        val root = trieDelete(trieRootOrBuild(), docId)
        val newEntries = entries - docId
        return DocumentTree(trieTreeHash(root), newEntries).also { it.trieRoot = root }
    }

    private fun trieRootOrBuild(): TrieNode? {
        val cached = trieRoot
        if (cached != null || entries.isEmpty()) return cached
        return trieBuild(entries)
    }

    public companion object {
        public val EMPTY: DocumentTree = build(emptyMap())

        /**
         * Builds a tree with content-addressed tree hash from a flat map,
         * building a full trie from scratch (O(n) - see trieBuild). Used
         * for trees not derived incrementally via with/without (e.g. wire
         * decode); the result still carries a trieRoot, so any subsequent
         * with/without on it is O(delta).
         */
        public fun build(entries: Map<KdbUuid, KdbHash>): DocumentTree {
            val root = trieBuild(entries)
            return DocumentTree(trieTreeHash(root), entries).also { it.trieRoot = root }
        }

        public fun fromKdbValue(value: KdbValue): DocumentTree {
            val arr = value as? KdbValue.ArrayVal ?: throw CommitDecodeException("DocumentTree: expected array")
            val m = linkedMapOf<KdbUuid, KdbHash>()
            for (e in arr.elements) {
                val r = e as? KdbValue.RecordVal ?: throw CommitDecodeException("DocumentTree: entry record")
                val id =
                    (r.fields[1] as? KdbValue.UuidVal)?.toKdbUuid()
                        ?: throw CommitDecodeException("DocumentTree: docId")
                val fh =
                    (r.fields[2] as? KdbValue.FixedVal)?.v
                        ?: throw CommitDecodeException("DocumentTree: contentHash")
                m[id] = KdbHash.fromBytes(fh.copyOf())
            }
            return build(m)
        }
    }
}

private fun entriesToArrayValue(entries: Map<KdbUuid, KdbHash>): KdbValue {
    // Sort keys are computed once per entry rather than via sortedBy's
    // Comparator (which calls the selector on every comparison, i.e.
    // O(n log n) toString() calls instead of O(n)): at 2000 entries this
    // was the dominant cost of BuildDocumentTree by ~two orders of
    // magnitude over the actual hash/encode work - see the Phase 3 note
    // in docs/benchmarks/phase0-baseline.md. Sort order (and therefore
    // the resulting hash) is unchanged.
    val sorted =
        entries.keys
            .map { it to it.toString() }
            .sortedBy { it.second }
            .map { it.first }
    return KdbValue.ArrayVal(
        sorted.map { id ->
            KdbValue.RecordVal(
                mapOf(
                    1 to id.toUuidVal(),
                    2 to KdbValue.FixedVal(entries.getValue(id).bytes.copyOf()),
                ),
            )
        },
    )
}

public fun DocumentTree.toKdbValue(): KdbValue = entriesToArrayValue(entries)

/** Named branch pointer. */
public data class KdbBranch(
    val name: String,
    val namespaceId: String,
    val headHash: KdbHash,
    val createdAt: KdbTimestamp,
    val updatedAt: KdbTimestamp,
)

/** Named tag (immutable ref). */
public data class KdbTag(
    val name: String,
    val namespaceId: String,
    val commitHash: KdbHash,
    val createdAt: KdbTimestamp,
    val message: String = "",
)

/** DAG placeholder for ice-archived commits. */
public data class CommitStub(
    val originalHash: KdbHash,
    val archiveLocation: String,
    val stubbedAt: KdbTimestamp,
) {
    public fun toKdbValue(): KdbValue =
        KdbValue.RecordVal(
            mapOf(
                1 to KdbValue.FixedVal(originalHash.bytes.copyOf()),
                2 to KdbValue.StringVal(archiveLocation),
                3 to stubbedAt.toTimestampVal(),
            ),
        )

    public companion object {
        public fun fromKdbValue(value: KdbValue): CommitStub {
            val r = value as? KdbValue.RecordVal ?: throw CommitDecodeException("CommitStub: expected record")
            val oh =
                (r.fields[1] as? KdbValue.FixedVal)?.v?.copyOf()
                    ?: throw CommitDecodeException("CommitStub: originalHash")
            val loc =
                (r.fields[2] as? KdbValue.StringVal)?.v
                    ?: throw CommitDecodeException("CommitStub: archiveLocation")
            val ts =
                (r.fields[3] as? KdbValue.TimestampVal)?.toKdbTimestamp()
                    ?: throw CommitDecodeException("CommitStub: stubbedAt")
            return CommitStub(KdbHash.fromBytes(oh), loc, ts)
        }
    }
}

File: go/kdb/document/document_tree.go
package document

import (
	"sort"

	"github.com/limidus/kdb/go/kdb/codec"
)

// DocumentTree is the doc id → content hash map at a commit.
//
// Entries is nil for a tree produced by With/Without - those are backed incrementally by
// trieRoot alone (see document_tree_trie.go) and deliberately do NOT eagerly materialize a flat
// map, which used to be copied in full on every single call (O(current document count), not
// O(delta)) and was found to be a major contributor to kdb-service getting OOM-killed under
// sustained write load once a namespace grew past a few hundred documents (see
// docs/benchmarks/lightsail-sim/README.md and go/kdb/transaction's
// BenchmarkCommitScalingWithHistorySize). Contains/HashFor/Size all work correctly regardless of
// whether Entries is populated (trieRoot-backed when it isn't); call MaterializedEntries()
// explicitly for the rare cases that genuinely need the full flat map (wire/storage
// serialization, DAG diff, full scans) - see its own doc comment.
//
// Entries is still populated for trees built directly from a flat map (BuildDocumentTree, e.g.
// wire decode) - no regression there, since that map already existed in full before construction.
type DocumentTree struct {
	TreeHash codec.Hash
	Entries  map[codec.UUID]codec.Hash

	// size backs Size() when Entries is nil (trie-backed tree) - tracked incrementally in
	// With/Without so Size() stays O(1) without needing a full trie walk.
	size int

	// trieRoot backs incremental With/Without updates (see
	// document_tree_trie.go): when present, TreeHash for a derived tree
	// is computed in O(delta) instead of O(len(Entries)). Trees built
	// via BuildDocumentTree from a flat map (e.g. wire decode) don't
	// carry one - With/Without fall back to a one-time O(n) trieBuild
	// in that case, then are incremental from then on. Always correct
	// either way; only the constant differs.
	trieRoot *trieNode
}

func (t DocumentTree) Size() int {
	if t.Entries != nil {
		return len(t.Entries)
	}
	return t.size
}

func (t DocumentTree) Contains(docID codec.UUID) bool {
	if t.Entries != nil {
		_, ok := t.Entries[docID]
		return ok
	}
	_, ok := trieGet(t.trieRoot, docID)
	return ok
}

func (t DocumentTree) HashFor(docID codec.UUID) (codec.Hash, bool) {
	if t.Entries != nil {
		h, ok := t.Entries[docID]
		return h, ok
	}
	return trieGet(t.trieRoot, docID)
}

// MaterializedEntries returns the full doc id → content hash map, building it from the trie
// (O(n)) if this tree doesn't already carry one. Only call this where a full map is genuinely
// needed (wire/storage serialization, DAG diff, full namespace scans) - never on a per-write
// hot path, which is exactly the mistake With/Without used to make on every single call.
func (t DocumentTree) MaterializedEntries() map[codec.UUID]codec.Hash {
	if t.Entries != nil {
		return t.Entries
	}
	return trieEntries(t.trieRoot)
}

func (t DocumentTree) With(docID codec.UUID, contentHash codec.Hash) (DocumentTree, error) {
	root := t.trieRootOrBuild()
	_, existed := trieGet(root, docID)
	newRoot := trieInsert(root, docID, contentHash)
	newSize := t.Size()
	if !existed {
		newSize++
	}
	return DocumentTree{TreeHash: trieTreeHash(newRoot), size: newSize, trieRoot: newRoot}, nil
}

func (t DocumentTree) Without(docID codec.UUID) (DocumentTree, error) {
	root := t.trieRootOrBuild()
	_, existed := trieGet(root, docID)
	newRoot := trieDelete(root, docID)
	newSize := t.Size()
	if existed {
		newSize--
	}
	return DocumentTree{TreeHash: trieTreeHash(newRoot), size: newSize, trieRoot: newRoot}, nil
}

// trieRootOrBuild returns t's trie root, building one from t.Entries (O(n))
// if t doesn't already carry one - e.g. a tree decoded from the wire via
// BuildDocumentTree/DocumentTreeFromValue. One-time cost; every
// subsequent With/Without on the result is O(delta).
func (t DocumentTree) trieRootOrBuild() *trieNode {
	if t.trieRoot != nil || t.Entries == nil || len(t.Entries) == 0 {
		return t.trieRoot
	}
	return trieBuild(t.Entries)
}

var emptyTree DocumentTree

func init() {
	var err error
	emptyTree, err = BuildDocumentTree(nil)
	if err != nil {
		panic(err)
	}
}

// EmptyDocumentTree is the canonical empty tree.
func EmptyDocumentTree() DocumentTree { return emptyTree }

// BuildDocumentTree constructs a tree with content-addressed tree hash,
// building a full trie from entries (O(n) - see trieBuild). Used for
// trees not derived incrementally via With/Without (e.g. wire decode);
// the resulting tree still carries a trie, so any subsequent With/Without
// on it is O(delta).
func BuildDocumentTree(entries map[codec.UUID]codec.Hash) (DocumentTree, error) {
	if entries == nil {
		entries = map[codec.UUID]codec.Hash{}
	}
	root := trieBuild(entries)
	return DocumentTree{TreeHash: trieTreeHash(root), Entries: entries, trieRoot: root}, nil
}

func entriesToArrayValue(entries map[codec.UUID]codec.Hash) codec.Value {
	// Sort keys are computed once per entry (O(n) String() calls) rather
	// than inside the comparator (O(n log n) calls, ~2x per comparison):
	// at 2000 entries this step alone was ~12ms, over 95% of
	// BuildDocumentTree's total cost and two orders of magnitude more
	// than the actual SHA256/wire-encode work - see the Phase 3 note in
	// docs/benchmarks/phase0-baseline.md. Sort order (and therefore the
	// resulting hash) is unchanged: same comparator, same strings.
	type keyed struct {
		id  codec.UUID
		key string
	}
	sorted := make([]keyed, 0, len(entries))
	for id := range entries {
		sorted = append(sorted, keyed{id: id, key: id.String()})
	}
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i].key < sorted[j].key
	})
	els := make([]codec.Value, len(sorted))
	for i, k := range sorted {
		els[i] = codec.RecordValue{Fields: map[int]codec.Value{
			1: uuidVal(k.id),
			2: hashVal(entries[k.id]),
		}}
	}
	return codec.ArrayValue{Elements: els}
}

// DocumentTreeFromValue decodes a wire document tree array.
func DocumentTreeFromValue(value codec.Value) (DocumentTree, error) {
	arr, ok := value.(codec.ArrayValue)
	if !ok {
		return DocumentTree{}, NewCommitDecodeError("DocumentTree: expected array", nil)
	}
	m := make(map[codec.UUID]codec.Hash, len(arr.Elements))
	for _, e := range arr.Elements {
		rec, ok := e.(codec.RecordValue)
		if !ok {
			return DocumentTree{}, NewCommitDecodeError("DocumentTree: entry record", nil)
		}
		id, err := uuidFromVal(rec.Fields[1])
		if err != nil {
			return DocumentTree{}, NewCommitDecodeError("DocumentTree: docId", nil)
		}
		h, err := hashFromVal(rec.Fields[2])
		if err != nil {
			return DocumentTree{}, err
		}
		m[id] = h
	}
	return BuildDocumentTree(m)
}

// Branch is a named branch pointer.
type Branch struct {
	Name        string
	NamespaceID string
	HeadHash    codec.Hash
	CreatedAt   codec.Timestamp
	UpdatedAt   codec.Timestamp
}

// Tag is a named immutable ref.
type Tag struct {
	Name        string
	NamespaceID string
	CommitHash  codec.Hash
	CreatedAt   codec.Timestamp
	Message     string
}

// CommitStub is a DAG placeholder for ice-archived commits.
type CommitStub struct {
	OriginalHash    codec.Hash
	ArchiveLocation string
	StubbedAt       codec.Timestamp
}

func (s CommitStub) ToValue() codec.Value {
	return codec.RecordValue{Fields: map[int]codec.Value{
		1: hashVal(s.OriginalHash),
		2: codec.StringValue{V: s.ArchiveLocation},
		3: timestampVal(s.StubbedAt),
	}}
}

// CommitStubFromValue decodes a wire commit stub record.
func CommitStubFromValue(value codec.Value) (CommitStub, error) {
	rec, ok := value.(codec.RecordValue)
	if !ok {
		return CommitStub{}, NewCommitDecodeError("CommitStub: expected record", nil)
	}
	oh, err := hashFromVal(rec.Fields[1])
	if err != nil {
		return CommitStub{}, NewCommitDecodeError("CommitStub: originalHash", nil)
	}
	loc, ok := rec.Fields[2].(codec.StringValue)
	if !ok {
		return CommitStub{}, NewCommitDecodeError("CommitStub: archiveLocation", nil)
	}
	ts, err := timestampFromVal(rec.Fields[3])
	if err != nil {
		return CommitStub{}, NewCommitDecodeError("CommitStub: stubbedAt", nil)
	}
	return CommitStub{OriginalHash: oh, ArchiveLocation: loc.V, StubbedAt: ts}, nil
}
